### PR TITLE
Allow binding to additional service instances.

### PR DESCRIPTION
**Pull request by jmcarp**
This is incomplete but ready for feedback. One thing I haven't figured out is how to cram the updated policy templates into `config-sample.json`, but maybe that means the idea of keeping go-templates that generate json inside a json file wasn't a good idea in the first place 🤔

### CODE REVIEW

File: broker/broker.go [hunk 1]
```diff
@@ -166,6 +179,42 @@ func (b *S3Broker) Deprovision(
...
+	}
+
+	instance, err := b.cfClient.ServiceInstanceByGuid(instanceGUID)
+	if err != nil {
+		return bucketNames, err
+	}
+
+	query := url.Values{}
+	query.Set("space_guid", instance.SpaceGuid)
+	query.Set("plan_guid", planGUID)
```
Comment (cnelson) [inline broker/broker.go]: Will this prevent you from making a credential that can access a basic and basic-public bucket?

Comment (jmcarp) [inline broker/broker.go]: Derp, yes, changing to service guid.

**Comment by jmcarp** [comment]
About templating: the problem is that the templates (which generate json) are themselves stored in a json file. So for example, at https://github.com/cloudfoundry-community/s3-broker/blob/master/config-sample.json#L80, we can't say

```json
...
"Resource": {{resources ""}}
...
```

Because it wouldn't be valid json. If we used yaml for the config, we could say

```yaml
iam_policy: |
  {
    ...
    "Resource": {{resources ""}
    ...
  }
```

because that's a valid inline string. Which is why I was thinking of changing the config to yaml, or allowing json and yaml for backwards compatibility. WDYT?

**Comment by cnelson** [comment]
JSON is YAML so no harm in making the switch?

File: awsiam/iam_user.go [hunk 2]
```diff
@@ -155,15 +156,24 @@ func (i *IAMUser) DeleteAccessKey(userName, accessKeyID string) error {
...
+			return string(marshaled)
+		},
+	}).Parse(policyTemplate)
 	if err != nil {
 		i.logger.Error("aws-iam-error", err)
 		return "", err
 	}
 	policy := bytes.Buffer{}
 	err = tmpl.Execute(&policy, map[string]interface{}{
-		"Resource": resource,
```
Comment (cnelson) [inline awsiam/iam_user.go]: We might want to still set resource to the first item in the slice for backwards compatibility?

File: CONFIGURATION.md [hunk 3]
```diff
@@ -1,6 +1,6 @@
 # Configuration
 
-A sample configuration can be found at [config-sample.json](https://github.com/cloudfoundry-community/s3-broker/blob/master/config-sample.json).
+A sample configuration can be found at [config-sample.yml](https://github.com/cloudfoundry-community/s3-broker/blob/master/config-sample.yml).
```
Comment (cnelson) [inline CONFIGURATION.md]: Rename / reformat the existing config-sample.json file as part of this?

**Comment by rogeruiz** [comment]
2k+ files changed.

plop

**Pull request merged**